File: test/webmention/rsslinkcollector.test.js

const { collect } = require('../../src/webmention/rsslinkcollector')
const fs = require('fs').promises
const dayjs = require('dayjs')

describe("collect RSS links of articles since certain period", () => {

	let xml = ''
	beforeEach(async () => {
		xml = (await fs.readFile('./test/__mocks__/samplerss.xml')).toString()
	})

	test("collect should not contain hrefs from blocked domains", () => {
		const collected = collect(xml, dayjs('2021-03-10T00:00:00.000Z').toDate())

		// test case: 
		// contains youtube.com/cool link
		const last = collected[collected.length - 1]
		expect(last.hrefs).toEqual([
			"https://dog.estate/@eli_oat",
			"https://twitter.com/olesovhcom/status/1369478732247932929",
			"/about"
		])

	})

	test("collect should not contain hrefs that point to images", () => {
		const collected = collect(xml, dayjs('2021-03-14T00:00:00.000Z').toDate())

		// test case: 
		// contains e.g. https://chat.brainbaking.com/media/6f8b72ca-9bfb-460b-9609-c4298a8cab2b/EuropeBattle%202021-03-14%2016-20-36-87.jpg
		const last = collected[collected.length - 1]
		expect(last.hrefs).toEqual([
			"/about"
		])
	})

	test("collects if time tag found in content that acts as an update stamp", async () => {
		// sample item: pubDate 2021-03-16, timestamp updated: 2021-03-20
		xml = (await fs.readFile('./test/__mocks__/samplerss-updated-timestamp.xml')).toString()

		const collected = collect(xml, dayjs('2021-03-19').toDate())
		expect(collected.length).toBe(1)
	})

	test("does not collect if time tag found in content but still older than since", async () => {
		// sample item: pubDate 2021-03-16, timestamp updated: 2021-03-20
		xml = (await fs.readFile('./test/__mocks__/samplerss-updated-timestamp.xml')).toString()

		const collected = collect(xml, dayjs('2021-03-21').toDate())
		expect(collected.length).toBe(0)
	})

	test("collects nothing if date in future and since nothing new in feed", () => {
		const collected = collect(xml, dayjs().add(7, 'day').toDate())
		expect(collected.length).toEqual(0)
	})

	test("collect latest x links when a since parameter is provided", () => {
		const collected = collect(xml, dayjs('2021-03-15T00:00:00.000Z').toDate())
		expect(collected.length).toEqual(3)

		const last = collected[collected.length - 1]
		expect(last.link).toBe("https://brainbaking.com/notes/2021/03/15h14m43s49/")
		expect(last.hrefs).toEqual([
			"http://replit.com",
			"http://codepen.io",
			"https://kuleuven-diepenbeek.github.io/osc-course/ch1-c/intro/",
			"/about"
		])
	})

	test("collect every external link without a valid since date", () => {
		const collected = collect(xml)
		expect(collected.length).toEqual(141)

		const first = collected[0]
		expect(first.link).toBe("https://brainbaking.com/notes/2021/03/16h17m07s14/")
		expect(first.hrefs).toEqual([
			"https://fosstodon.org/@celia",
			"https://fosstodon.org/@kev",
			"/about"
		])
	})

})

File: app/webmention/send/rsslinkcollector_test.go
package send

import (
	"brainbaking.com/go-jamming/common"
	"github.com/stretchr/testify/assert"
	"io/ioutil"
	"testing"
)

func TestCollectUniqueHrefsFromHtml(t *testing.T) {
	cases := []struct {
		label         string
		html          string
		expectedLinks []string
	}{
		{
			"should not contain inline links",
			`<html><body><a href="#inline">sup</a></body></html>`,
			[]string{},
		},
		{
			"should not collect blacklisted links",
			`<html><body><a href="https://www.blacklisted.com/wowo.html">sup</a> and also <a href="/dinges">dinges</a>!</body></html>`,
			[]string{
				"/dinges",
			},
		},
		{
			"should not collect hrefs from <link/> tags, only from <a/> ones",
			`<html><head><link rel="stylesheet" href="/style.css"></head><body><a href="/dinges">dinges</a>!</body></html>`,
			[]string{
				"/dinges",
			},
		},
		{
			"should collect even if href is not the first attribute of an <a> tag",
			`<html><body><a style="cool" target="_blank" href="/one">one</a> and <a target="_blank" href="/two">two</a> and <a href="/three">three</a></body></html>`,
			[]string{
				"/one",
				"/two",
				"/three",
			},
		},
		{
			"should collect case insensitive",
			`<html><body><A href="/one">one</A> and <a href="/two">two</a> and <a HREF="/three">three</a></body></html>`,
			[]string{
				"/one",
				"/two",
				"/three",
			},
		},
		{
			"should not collect zips or ZIPs or gifs or GIFS",
			`<a href="/cool.gif">cool gif</a> and <a href="/more-cool.GIF">more-cool gif</a> and here's a zip: <a href="baf.ZIP">baf</a> or <a href="boef.zip">boef.zip</a>??'`,
			[]string{},
		},
	}

	s := &Sender{
		Conf: &common.Config{
			Blacklist: []string{
				"blacklisted.com",
			},
		},
	}
	for _, tc := range cases {
		t.Run(tc.label, func(t *testing.T) {
			result := s.collectUniqueHrefsFromHtml(tc.html)
			assert.ElementsMatch(t, tc.expectedLinks, result)
		})
	}
}

func TestCollect(t *testing.T) {
	file, _ := ioutil.ReadFile("../../../mocks/samplerss.xml")
	snder := &Sender{
		Conf: &common.Config{
			Blacklist: []string{
				"youtube.com",
			},
		},
	}

	cases := []struct {
		label             string
		lastsentlink      string
		expectedRssItems  int
		expectedLastLinks []string
	}{
		{
			"should not contain hrefs from blocked domains",
			"https://brainbaking.com/notes/2021/03/09h15m17s30/",
			10,
			[]string{
				"https://dog.estate/@eli_oat",
				"https://twitter.com/olesovhcom/status/1369478732247932929",
				"/about",
			},
		},
		{
			// test case: contains e.g. https://chat.brainbaking.com/media/6f8b72ca-9bfb-460b-9609-c4298a8cab2b/EuropeBattle%202021-03-14%2016-20-36-87.jpg
			"should not contain hrefs that point to images",
			"https://brainbaking.com/notes/2021/03/13h12m44s29/",
			4,
			[]string{
				"/about",
			},
		},
		{
			"collects nothing if nothing new in feed",
			"https://brainbaking.com/notes/2021/03/16h17m07s14/",
			0,
			[]string{},
		},
		{
			"collect latest X links when a recent link parameter is provided",
			"https://brainbaking.com/notes/2021/03/14h17m41s53/",
			3,
			[]string{
				"http://replit.com",
				"http://codepen.io",
				"https://kuleuven-diepenbeek.github.io/osc-course/ch1-c/intro/",
				"/about",
			},
		},
		{
			"collect every external link without a recent link",
			"",
			141,
			[]string{
				"/notes/index.xml",
				"/archives",
				"/categories/hardware/index.xml",
				"/about",
				"https://netnewswire.com/",
				"/index.xml",
				"brainbaking.com",
				"/post/index.xml",
			},
		},
	}

	for _, tc := range cases {
		t.Run(tc.label, func(t *testing.T) {
			items, err := snder.Collect(string(file), tc.lastsentlink)
			assert.NoError(t, err)
			assert.Equal(t, tc.expectedRssItems, len(items))

			if tc.expectedRssItems > 0 {
				last := items[len(items)-1]
				assert.ElementsMatch(t, tc.expectedLastLinks, last.hrefs)
			}
		})
	}
}
